### ISSUE TITLE
Add field additional uuid

### ISSUE THREAD

**Issue by franviera92** [JosephSilber/bouncer#397]
Hi guys.!

I have the following migration:

Schema::create('users', function (Blueprint $table) {
            $table->increments('id');
            $table->uuid('uuid');
            $table->string('username',50)->unique();
            $table->string('identity_card',20)->unique();
            $table->string('first_name',60);
            $table->string('second_name',60)->nullable();
            $table->string('last_name',60);
            $table->string('second_last_name',60)->nullable();
            $table->string('name');
            $table->string('email')->unique();
            $table->string('password');
            $table->rememberToken();
            $table->timestamps();
        });

What can I do so that my entity_id recognizes the new field uuid?

I don't work with the field id :/

example with the table permisssions?

**Comment by PepijnDeWachter** [comment]
Follow the docs to publish the migration file and change it in your own migrations folder.
I'm using uuid for everything so I also made my own ability and role models that use uuids, if you only use uuid on the user you will only need to change the entity_id to uuid.
When using roles on user you'll have to change it on the assigned_roles table aswel
If you never give permissions directly to a user, only permissions on roles and roles on users, then you shouldn't need to edit the permissions table, since entity_id would always be an int from the 'role_id'.
However if a permission entity_id could be both from a role and a user (when you give abilities to both users and roles) then you will have to change the id on the roles table to a uuid as well, and you would have to extend the Role model as explained in the docs.
permissions:
 `Schema::create(Models::table('permissions'), function (Blueprint $table) {
            $table->increments('id');
            $table->uuid('ability_id')->index();
            $table->uuid('entity_id')->nullable();
            $table->string('entity_type')->nullable();
            $table->boolean('forbidden')->default(false);
            $table->integer('scope')->nullable()->index();

            $table->index(
                ['entity_id', 'entity_type', 'scope'],
                'permissions_entity_index'
            );

            $table->foreign('ability_id')
                  ->references('id')->on(Models::table('abilities'))
                  ->onUpdate('cascade')->onDelete('cascade');
        });`

assigned_roles: 
`Schema::create(Models::table('assigned_roles'), function (Blueprint $table) {
            $table->increments('id');
            $table->uuid('role_id')->index();
            $table->uuid('entity_id');
            $table->string('entity_type');
            $table->uuid('restricted_to_id')->nullable();
            $table->string('restricted_to_type')->nullable();
            $table->integer('scope')->nullable()->index();

            $table->index(
                ['entity_id', 'entity_type', 'scope'],
                'assigned_roles_entity_index'
            );

            $table->foreign('role_id')
                  ->references('id')->on(Models::table('roles'))
                  ->onUpdate('cascade')->onDelete('cascade');
        });`

I hope this helps you

**Comment by franviera92** [comment]
> Follow the docs to publish the migration file and change it in your own migrations folder.
> I'm using uuid for everything so I also made my own ability and role models that use uuids, if you only use uuid on the user you will only need to change the entity_id to uuid.
> When using roles on user you'll have to change it on the assigned_roles table aswel
> If you never give permissions directly to a user, only permissions on roles and roles on users, then you shouldn't need to edit the permissions table, since entity_id would always be an int from the 'role_id'.
> However if a permission entity_id could be both from a role and a user (when you give abilities to both users and roles) then you will have to change the id on the roles table to a uuid as well, and you would have to extend the Role model as explained in the docs.
> permissions:
> 
> ```
> Schema::create(Models::table('permissions'), function (Blueprint $table) {
>             $table->increments('id');
>             $table->uuid('ability_id')->index();
>             $table->uuid('entity_id')->nullable();
>             $table->string('entity_type')->nullable();
>             $table->boolean('forbidden')->default(false);
>             $table->integer('scope')->nullable()->index();
> 
>             $table->index(
>                 ['entity_id', 'entity_type', 'scope'],
>                 'permissions_entity_index'
>             );
> 
>             $table->foreign('ability_id')
>                   ->references('id')->on(Models::table('abilities'))
>                   ->onUpdate('cascade')->onDelete('cascade');
>         });
> ```
> assigned_roles:
> 
> ```
> Schema::create(Models::table('assigned_roles'), function (Blueprint $table) {
>             $table->increments('id');
>             $table->uuid('role_id')->index();
>             $table->uuid('entity_id');
>             $table->string('entity_type');
>             $table->uuid('restricted_to_id')->nullable();
>             $table->string('restricted_to_type')->nullable();
>             $table->integer('scope')->nullable()->index();
> 
>             $table->index(
>                 ['entity_id', 'entity_type', 'scope'],
>                 'assigned_roles_entity_index'
>             );
> 
>             $table->foreign('role_id')
>                   ->references('id')->on(Models::table('roles'))
>                   ->onUpdate('cascade')->onDelete('cascade');
>         });
> ```
> I hope this helps you

Hello @PepijnDeWachter , 

It is very interesting what you mention. At this moment I only use uuid in the Users table, but as I said before, my user table contains 1 field ID and 1 UUID field, besides modifying the bouncer migrations $ table-> uuid ('entity_id'), I have What else should I do for the entity_id field to use my new field uuid from the user table with the coexistence of the id and uuid fields in the user table?

**Comment by PepijnDeWachter** [comment]
> Hello @PepijnDeWachter ,
> 
> It is very interesting what you mention. At this moment I only use uuid in the Users table, but as I said before, my user table contains 1 field ID and 1 UUID field, besides modifying the bouncer migrations $ table-> uuid ('entity_id'), I have What else should I do for the entity_id field to use my new field uuid from the user table with the coexistence of the id and uuid fields in the user table?

Bouncer will use the result of the getKey() method on your User model to decide wich field to use for the relationship.
Default Laravel returns 'id' unless you override that method in your model, however usually when working with a UUID as the main id you would remove the original id on your table since it serves no purpose, you could then use the creating eloquent event hook to automatically generate a new uuid for the user when creating one, I've had a few cases where I had both a uuid and an auto increment column in the same table but that auto increment column was always for something else and never the primarykey of that table.

Basically there is almost no good reason for having 2 id fields on a table, unless you're migrating from an autoincrement old database to a 'secure' uuid based database
if you really don't use the incrementing id field anywhere, I suggest just getting rid of it, and naming the uuid field 'id'

**Comment by franviera92** [comment]
@PepijnDeWachter 

If you are right, I have agreed to use only ID for the Uuid.

Schema::create(Models::table('abilities'), function (Blueprint $table) {
            $table->uuid('id')->primary();
            $table->string('name');
            $table->string('title')->nullable();
            $table->uuid('entity_id')->nullable();
            $table->string('entity_type')->nullable();
            $table->boolean('only_owned')->default(false);
            $table->json('options')->nullable();
            $table->integer('scope')->nullable()->index();
            $table->string('parent')->nullable();
            $table->string('icon')->nullable();
            $table->timestamps();
        });

        Schema::create(Models::table('roles'), function (Blueprint $table) {
            $table->uuid('id')->primary();
            $table->string('name');
            $table->string('title')->nullable();
            $table->integer('level')->unsigned()->nullable();
            $table->integer('scope')->nullable()->index();
            $table->timestamps();

            $table->unique(
                ['name', 'scope'],
                'roles_name_unique'
            );
        });

        Schema::create(Models::table('assigned_roles'), function (Blueprint $table) {
            $table->uuid('role_id')->index();
            $table->uuid('entity_id')->nullable();
            $table->string('entity_type');
            $table->uuid('restricted_to_id')->nullable();
            $table->string('restricted_to_type')->nullable();
            $table->integer('scope')->nullable()->index();

            $table->index(
                ['entity_id', 'entity_type', 'scope'],
                'assigned_roles_entity_index'
            );

            $table->foreign('role_id')
                  ->references('id')->on(Models::table('roles'))
                  ->onUpdate('cascade')->onDelete('cascade');
        });

        Schema::create(Models::table('permissions'), function (Blueprint $table) {
            $table->uuid('ability_id')->index();
            $table->uuid('entity_id')->nullable();
            $table->string('entity_type');
            $table->boolean('forbidden')->default(false);
            $table->integer('scope')->nullable()->index();

            $table->index(
                ['entity_id', 'entity_type', 'scope'],
                'permissions_entity_index'
            );

            $table->foreign('ability_id')
                  ->references('id')->on(Models::table('abilities'))
                  ->onUpdate('cascade')->onDelete('cascade');
        });

Now I want to add a custom method for the ability and Role model.

I have created the following Models:

ABILITY

<? php

namespace App;

use Illuminate \ Support \ Collection;
use Illuminate \ Database \ Eloquent \ Model;
use Silber \ Bouncer \ Database \ Ability;

class CustomsAbility extends Ability
{

    protected $ table = 'abilities';

    public $ incrementing = false;

    protected $ keyType = 'string';
}

namespace App;

use Illuminate \ Support \ Collection;
use Illuminate \ Database \ Eloquent \ Model ;;
use Silber \ Bouncer \ Database \ Role;

class CustomRole extends Role
{
    

    protected $ table = 'roles';

    public $ incrementing = false;

    protected $ keyType = 'string';
    
}


How can I make use of these new models?

The documentation makes clear that it is with this:
 Bouncer :: useAbilityModel (MyAbility :: class);
Bouncer :: useRoleModel (MyRole :: class);

But where should I register this in my models or in the controllers?

I would like to use for example the following in my controller

use Silber \ Bouncer \ Database \ Ability;
use Silber \ Bouncer \ Database \ Role;

Method store
$ user-> allow ($ abilitiesUsers, User :: class);

**Comment by franviera92** [comment]
I have created the following Models:

ABILITY

<? php

namespace App;

use Illuminate \ Support \ Collection;
use Illuminate \ Database \ Eloquent \ Model;
use Silber \ Bouncer \ Database \ Ability;

class CustomsAbility extends Ability
{

    protected $ table = 'abilities';

    public $ incrementing = false;

    protected $ keyType = 'string';
}

namespace App;

use Illuminate \ Support \ Collection;
use Illuminate \ Database \ Eloquent \ Model ;;
use Silber \ Bouncer \ Database \ Role;

class CustomRole extends Role
{
    

    protected $ table = 'roles';

    public $ incrementing = false;

    protected $ keyType = 'string';
    
}


How can I make use of these new models?

The documentation makes clear that it is with this:
 Bouncer :: useAbilityModel (MyAbility :: class);
Bouncer :: useRoleModel (MyRole :: class);

But where should I register this in my models or in the controllers?

I would like to use for example the following in my controller

use Silber \ Bouncer \ Database \ Ability;
use Silber \ Bouncer \ Database \ Role;

Method store
$ user-> allow ($ abilitiesUsers, User :: class);

**Comment by franviera92** [comment]
@PepijnDeWachter 

Now I want to add a custom method for the ability and Role model.

Schema::create(Models::table('abilities'), function (Blueprint $table) {
            $table->uuid('id')->primary();
            $table->string('name');
            $table->string('title')->nullable();
            $table->uuid('entity_id')->nullable();
            $table->string('entity_type')->nullable();
            $table->boolean('only_owned')->default(false);
            $table->json('options')->nullable();
            $table->integer('scope')->nullable()->index();
            $table->string('parent')->nullable();
            $table->string('icon')->nullable();
            $table->timestamps();
        });

        Schema::create(Models::table('roles'), function (Blueprint $table) {
            $table->uuid('id')->primary();
            $table->string('name');
            $table->string('title')->nullable();
            $table->integer('level')->unsigned()->nullable();
            $table->integer('scope')->nullable()->index();
            $table->timestamps();

            $table->unique(
                ['name', 'scope'],
                'roles_name_unique'
            );
        });

        Schema::create(Models::table('assigned_roles'), function (Blueprint $table) {
            $table->uuid('role_id')->index();
            $table->uuid('entity_id')->nullable();
            $table->string('entity_type');
            $table->uuid('restricted_to_id')->nullable();
            $table->string('restricted_to_type')->nullable();
            $table->integer('scope')->nullable()->index();

            $table->index(
                ['entity_id', 'entity_type', 'scope'],
                'assigned_roles_entity_index'
            );

            $table->foreign('role_id')
                  ->references('id')->on(Models::table('roles'))
                  ->onUpdate('cascade')->onDelete('cascade');
        });

        Schema::create(Models::table('permissions'), function (Blueprint $table) {
            $table->uuid('ability_id')->index();
            $table->uuid('entity_id')->nullable();
            $table->string('entity_type');
            $table->boolean('forbidden')->default(false);
            $table->integer('scope')->nullable()->index();

            $table->index(
                ['entity_id', 'entity_type', 'scope'],
                'permissions_entity_index'
            );

            $table->foreign('ability_id')
                  ->references('id')->on(Models::table('abilities'))
                  ->onUpdate('cascade')->onDelete('cascade');
        });

**Comment by franviera92** [comment]
I have created the following Models:

ABILITY

<? php

namespace App;

use Illuminate \ Support \ Collection;
use Illuminate \ Database \ Eloquent \ Model;
use Silber \ Bouncer \ Database \ Ability;

class CustomsAbility extends Ability
{

    protected $ table = 'abilities';

    public $ incrementing = false;

    protected $ keyType = 'string';
}

namespace App;

use Illuminate \ Support \ Collection;
use Illuminate \ Database \ Eloquent \ Model ;;
use Silber \ Bouncer \ Database \ Role;

class CustomRole extends Role
{
    

    protected $ table = 'roles';

    public $ incrementing = false;

    protected $ keyType = 'string';
    
}


How can I make use of these new models?

The documentation makes clear that it is with this:
 Bouncer :: useAbilityModel (MyAbility :: class);
Bouncer :: useRoleModel (MyRole :: class);

But where should I register this in my models or in the controllers?

I would like to use for example the following in my controller

use Silber \ Bouncer \ Database \ Ability;
use Silber \ Bouncer \ Database \ Role;

Method store
$ user-> allow ($ abilitiesUsers, User :: class);

**Comment by franviera92** [comment]
@PepijnDeWachter 
I have created the following Models:
```


ABILITY

<? php

namespace App;

use Illuminate \ Support \ Collection;
use Illuminate \ Database \ Eloquent \ Model;
use Silber \ Bouncer \ Database \ Ability;

class CustomsAbility extends Ability
{

    protected $ table = 'abilities';

    public $ incrementing = false;

    protected $ keyType = 'string';
}

namespace App;

use Illuminate \ Support \ Collection;
use Illuminate \ Database \ Eloquent \ Model ;;
use Silber \ Bouncer \ Database \ Role;

class CustomRole extends Role
{
    

    protected $ table = 'roles';

    public $ incrementing = false;

    protected $ keyType = 'string';
    
}


How can I make use of these new models?

The documentation makes clear that it is with this:
 Bouncer :: useAbilityModel (MyAbility :: class);
Bouncer :: useRoleModel (MyRole :: class);

But where should I register this in my models or in the controllers?

I would like to use for example the following in my controller

use Silber \ Bouncer \ Database \ Ability;
use Silber \ Bouncer \ Database \ Role;

Method store
$ user-> allow ($ abilitiesUsers, User :: class);


```

**Comment by PepijnDeWachter** [comment]
> The documentation makes clear that it is with this:
>  Bouncer :: useAbilityModel (MyAbility :: class);
> Bouncer :: useRoleModel (MyRole :: class);

As noted in the documentation you should put those method calls in a service provider, if you only use a few configuration calls you can do this in the boot method of your AppServiceProvider.php
But if like me you have an application where you also plan to add a bunch of custom ownership rules (ownedvia, as explained in the documentation) you should create your own serviceprovider to put them in. This process is explained in the documentation, should come up easily with a search.

**Comment by franviera92** [comment]
Thank you, implement successfully uuid in conjunction with bouncer.

**Comment by Devbymitch** [comment]
For posterity: I'm currently converting a system to UUIDs for a distributed environment. `permissions` and `assigned_roles` both have primary auto-incrementing keys added to [fix this Nova issue someone was having](https://github.com/JosephSilber/bouncer/issues/382). We use Laravel < 6, so rc7 or higher is moot. [examining the file dif between rc.4 and rc.6](https://github.com/JosephSilber/bouncer/compare/v1.0.0-rc.4...v1.0.0-rc.6) didn't show any places I could find where these new keys were used at all, so I just migrated dropping `permissions.id` and `assigned_roles.id`.

Seems to be working so far, and if we ever use Nova for this project (blegh), we'd have to figure something out. If you do use Nova, it probably wouldn't be too much work to PR something with an option for setting pivot table primary ID type to UUID, but this works for us!
